cogs/roles.py

```diff
@@ -0,0 +1,166 @@
+# under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Canary is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Canary. If not, see <https://www.gnu.org/licenses/>.
+
+# discord-py requirements
+import discord
+from discord import utils
+from discord.ext import commands
+
+import codecs
+import configparser
+from .utils.paginator import Pages
+
+
+class Roles(commands.Cog):
+    def __init__(self, bot):
+        self.bot = bot
+        self.roles = self.bot.config.roles
+
+    async def toggle_role(self, ctx, transaction, requested_role, category):
+        """
+        Assigns a single role to a user with no checks from a category of roles
+        """
+        member = ctx.message.author
+        roles = self.roles[category]
+
+        # normalize user input to title case
+        requested_role = requested_role.title()
+        role = discord.utils.get(ctx.guild.roles, name=requested_role)
+
+        if (requested_role not in roles):
+            await ctx.send("Invalid Role")
+            return
+
+        if (transaction == "add"):
+            await member.add_roles(role, reason="Self Requested")
+        elif (transaction == "remove"):
+            await member.remove_roles(role, reason="Self Requested")
+        else:
+            await ctx.send("Must `add` or `remove` a role")
+
+    async def toggle_unique_role(self, ctx, transaction, requested_role,
+                                 category):
+        """
+        Assigs a single role to a user with no checks from a category of roles
+        """
+        member = ctx.message.author
+        roles = self.roles[category]
+
+        # normalize user input to title case
+        requested_role = requested_role.title()
+        role = discord.utils.get(ctx.guild.roles, name=requested_role)
+
+        if (requested_role not in roles):
+            await ctx.send("Invalid Role")
+            return
+
+        if (transaction == "add"):
+            # Allow only a single faculty role per user.
+            for category_roles in roles:
+                old_role = discord.utils.get(ctx.guild.roles,
+                                             name=category_roles)
+                if old_role is not None:
+                    await member.remove_roles(old_role,
+                                              reason="Self Requested")
+            await member.add_roles(role, reason="Self Requested")
+        elif (transaction == "remove"):
+            await member.remove_roles(role, reason="Self Requested")
+        else:
+            await ctx.send("Must `add` or `remove` a role")
```

Comment:
I see a lot of duplicated code between these 2 functions – is it possible to factor out the common parts, and functionalize the "precluding" operation in `toggle_unique_role()`?

Another option would be to write a general function for role addition and add an enclosing category around the faculty and year categories to denote them as single-assignment, then include a check in the function to see if the category argument is in the "single-assignment" category.